WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Helpers/DashboardPostsSyncManager.swift

```diff
@@ -0,0 +1,108 @@
+import Foundation
+
+protocol DashboardPostsSyncManagerListener: AnyObject {
+    func postsSynced(success: Bool, blog: Blog, posts: [AbstractPost]?, for statuses: [String])
+}
+
+class DashboardPostsSyncManager {
+
+    // MARK: Type Aliases
+
+    typealias SyncSuccessBlock = () -> Void
+    typealias SyncFailureBlock = (Error?) -> Void
+
+    // MARK: Private Variables
+
+    private let postService: PostService
+    private let blogService: BlogService
+    private let lock = NSRecursiveLock()
+    @Atomic private var listeners: [DashboardPostsSyncManagerListener] = []
+
+    // MARK: Shared Instance
+
+    static let shared = DashboardPostsSyncManager()
+
+    // MARK: Initializer
+
+    init(postService: PostService = PostService(managedObjectContext: ContextManager.shared.mainContext),
+         blogService: BlogService = BlogService(managedObjectContext: ContextManager.shared.mainContext)) {
+        self.postService = postService
+        self.blogService = blogService
+    }
+
+    // MARK: Public Functions
+
+    func addListener(_ listener: DashboardPostsSyncManagerListener) {
+        listeners.append(listener)
+    }
+
+    func removeListener(_ listener: DashboardPostsSyncManagerListener) {
+        if let index = listeners.firstIndex(where: {$0 === listener}) {
+            listeners.remove(at: index)
+        }
+    }
+
+    func syncPosts(blog: Blog, statuses: [String]) {
+        let toBeSynced = statusesNotBeingSynced(statuses, for: blog)
+        guard toBeSynced.isEmpty == false else {
+            return
+        }
+
+        blog.dashboardState.syncingStatuses.append(contentsOf: toBeSynced)
+
+        let options = PostServiceSyncOptions()
+        options.statuses = toBeSynced
+        options.authorID = blog.userID
+        options.number = Constants.numberOfPostsToSync
+        options.order = .descending
+        options.orderBy = .byModified
+        options.purgesLocalSync = true
+
+        // If the userID is nil we need to sync authors
+        // But only if the user is an admin
+        if blog.userID == nil && blog.isAdmin {
+            syncAuthors(blog: blog, success: { [weak self] in
+                self?.stopSyncingStatuses(toBeSynced, for: blog)
+                self?.syncPosts(blog: blog, statuses: toBeSynced)
+            }, failure: { [weak self] error in
+                self?.stopSyncingStatuses(toBeSynced, for: blog)
+                self?.notifyListenersOfPostsSync(success: false, blog: blog, posts: nil, for: toBeSynced)
+            })
+            return
+        }
+
+        postService.syncPosts(ofType: .post, with: options, for: blog) { [weak self] posts in
```

Comment:
It's way better to do one call than multiple but this logic has an issue: it requests 3 posts, ordered by modified with the given statuses.

It works fine when it's just draft or scheduled, but if the user has scheduled and drafts it won't return the specific posts we need (3 more recent scheduled and 3 more recent drafts). You can test this by:

1. Have a site with 4 or more posts, being 3 drafts and 1 scheduled or 2 drafts and 3 scheduled
2. Fresh install
3. Login
4. Go to the dashboard
5. Only the 3 most recent posts will be shown

Unfortunately, I don't think the API can handle this case, so we need to make two requests. ¶

¶ Ideally the dashboard API would already return the post — however that proved to be complex. :(

Reply:
Nice catch @leandroalonso!
Fixed in 5883e4fb8ca6a085df279f52857b1095558b3705